lib/AST/ClangTypeConverter.cpp

```diff
@@ -0,0 +1,728 @@
+          auto NSUIntegerTy = getClangBuiltinTypeFromTypedef(sema, "NSUInteger");
+          if (!NSUIntegerTy.isNull()) {
+            Cache.insert({swiftType->getCanonicalType(), NSUIntegerTy});
+            return;
+          }
+        }
+      }
+
+      // For something like `typealias CInt = Int32`, reverseBuiltinTypeMapping
+      // will get Int32 as the input, so we need to record the desugared type.
+      Cache.insert({swiftType->getCanonicalType(),
+                    getClangBuiltinTypeFromKind(ctx, builtinKind)});
+    };
+
+#define MAP_BUILTIN_TYPE(CLANG_BUILTIN_KIND, SWIFT_TYPE_NAME)          \
+    cacheStdlibType(#SWIFT_TYPE_NAME, clang::BuiltinType::CLANG_BUILTIN_KIND);
+#include "swift/ClangImporter/BuiltinMappedTypes.def"
+
+    // On 64-bit Windows, no C type is imported as an Int or UInt; CLong is
+    // imported as an Int32 and CLongLong as an Int64. Therefore, manually
+    // add mappings to C for Int and UInt.
+    // On 64-bit Cygwin, no manual mapping is required.
+    if (Triple.isOSWindows() && Triple.isArch64Bit()
+        && !Triple.isWindowsCygwinEnvironment()) {
+      // Map UInt to uintptr_t
+      auto swiftUIntType = getNamedSwiftType(stdlib, "UInt");
+      auto clangUIntPtrType = ctx.getCanonicalType(ctx.getUIntPtrType());
+      Cache.insert({swiftUIntType, clangUIntPtrType});
+
+      // Map Int to intptr_t
+      auto swiftIntType = getNamedSwiftType(stdlib, "Int");
+      auto clangIntPtrType = ctx.getCanonicalType(ctx.getIntPtrType());
+      Cache.insert({swiftIntType, clangIntPtrType});
+    }
+    StdlibTypesAreCached = true;
+  }
+
+  auto it = Cache.find(Type(type));
+  if (it != Cache.end())
+    return it->second;
+
+  it = Cache.find(type->getCanonicalType());
+  if (it != Cache.end()) {
+    Cache.insert({Type(type), it->second});
+    return it->second;
+  }
+
+  return clang::QualType();
+}
+
+clang::QualType ClangTypeConverter::visitTupleType(TupleType *type) {
+  unsigned tupleNumElements = type->getNumElements();
+  if (tupleNumElements == 0)
+    return ClangASTContext.VoidTy;
+
+  Type eltTy = type->getElementType(0);
+  for (unsigned i = 1; i < tupleNumElements; i++) {
+    if (!eltTy->isEqual(type->getElementType(i)))
+      // Only tuples where all element types are equal map to fixed-size
+      // arrays.
+      return clang::QualType();
+  }
+
+  auto clangEltTy = convert(eltTy);
+  if (clangEltTy.isNull())
+    return clang::QualType();
+
+  APInt size(32, tupleNumElements);
+  return ClangASTContext.getConstantArrayType(clangEltTy, size,
+           clang::ArrayType::Normal, 0);
+}
+
+clang::QualType ClangTypeConverter::visitProtocolType(ProtocolType *type) {
+  auto proto = type->getDecl();
+  auto &clangCtx = ClangASTContext;
+
+  if (!proto->isObjC())
+    return clang::QualType();
+
+  // TODO: As an optimization, we could check if proto has a clang decl
+  // already attached to it and if so, reuse it instead of making a new one.
```

Comment:
The problem is the const-ness. You mentioned earlier that other things would need to be refactored to avoid a `const_cast` here. I didn't want to do that refactoring and I didn't want to do that `const_cast` either.

Reply:
Ah never mind, we can replace this TODO with an assert because of the optimization.